### PR TITLE
Implement all supported thermocouple types for MAX31856

### PR DESCRIPTION
**Pull request by ArkanStasarik**
# What does this implement/fix?

MAX31856 supports 8 different thermocouple types however this was not implement. This implements all thermocouple types supported by the sensor

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4119

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/max31856/max31856.h [hunk 1]
```diff
@@ -79,10 +89,14 @@ class MAX31856Sensor : public sensor::Sensor,
   void dump_config() override;
   float get_setup_priority() const override;
   void set_filter(MAX31856ConfigFilter filter) { filter_ = filter; }
+  void set_thermocouple_type(MAX31856ConfigThermocoupleType thermocouple_type) {
+    thermocouple_type_ = thermocouple_type;
```
Comment (jesserockz) [inline esphome/components/max31856/max31856.h]:
```suggestion
  void set_filter(MAX31856ConfigFilter filter) { this->filter_ = filter; }
  void set_thermocouple_type(MAX31856ConfigThermocoupleType thermocouple_type) {
    this->thermocouple_type_ = thermocouple_type;
```

File: esphome/components/max31856/max31856.h [hunk 2]
```diff
@@ -70,6 +69,17 @@ enum MAX31856ConfigFilter {
...
+enum MAX31856ConfigThermocoupleType {
+  TYPE_B = 0,
+  TYPE_E = 1,
+  TYPE_J = 2,
+  TYPE_K = 3,
+  TYPE_N = 4,
+  TYPE_R = 5,
+  TYPE_S = 6,
+  TYPE_T = 7,
+};
```
Comment (jesserockz) [inline esphome/components/max31856/max31856.h]: You might as well reuse the `MAX31856ThermocoupleType` enum instead of making another one?

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/max31856/max31856.cpp [hunk 3]
```diff
@@ -32,6 +32,26 @@ void MAX31856Sensor::dump_config() {
...
+  } else if (thermocouple_type_ == TYPE_N) {
+    ESP_LOGCONFIG(TAG, "  Thermocouple Type: N");
+  } else if (thermocouple_type_ == TYPE_R) {
+    ESP_LOGCONFIG(TAG, "  Thermocouple Type: R");
+  } else if (thermocouple_type_ == TYPE_S) {
+    ESP_LOGCONFIG(TAG, "  Thermocouple Type: S");
+  } else if (thermocouple_type_ == TYPE_T) {
+    ESP_LOGCONFIG(TAG, "  Thermocouple Type: T");
+  } else {
+    ESP_LOGCONFIG(TAG, "  Thermocouple Type: Unknown");
```
Comment (clydebarrow) [inline esphome/components/max31856/max31856.cpp]:
```suggestion
  ESP_LOGCONFIG(TAG, "  Thermocouple Type: %c", "BEJKNRST"[this->thermocouple_type_]);
```
You could check for it being in range, but that's guaranteed by the config anyway. Or use a `std::string` and the `at` method, which bounds-checks.

File: esphome/components/max31856/max31856.h (hunk 2 above)
Comment (ArkanStasarik) [reply esphome/components/max31856/max31856.h]: You are right! Fixed now

**Review by jesserockz — APPROVED** [review]

**Pull request merged**